Add autograd not implemented boxed fallback

soulitzer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #63878
* #63459
* __->__ #63458

See description and discussion from https://github.com/pytorch/pytorch/pull/62450

Differential Revision: [D30518572](https://our.internmc.facebook.com/intern/diff/D30518572)

soulitzer committed:
Add autograd not implemented boxed fallback

[ghstack-poisoned]

soulitzer committed:
Update on "Add autograd not implemented boxed fallback"




[ghstack-poisoned]

soulitzer committed:
Update on "Add autograd not implemented boxed fallback"




[ghstack-poisoned]

soulitzer commented on torch/csrc/autograd/function.h:
```diff
@@ -563,6 +563,14 @@ struct MakeNextFunctionList : IterArgs<MakeNextFunctionList> {
       next_edges.emplace_back();
     }
   }
+  void operator()(const Variable* variable) {
```
Not sure how to remove this overload yet.

soulitzer committed:
Update on "Add autograd not implemented boxed fallback"


See description and discussion from https://github.com/pytorch/pytorch/pull/62450

[ghstack-poisoned]

ezyang commented on torch/csrc/autograd/autograd_not_implemented_fallback.cpp:
```diff
@@ -0,0 +1,180 @@
+    size_t size) {
+  // Enumerate over tensors in a stack, including ones in TensorLists
+  int idx = 0;
+  for (const auto i : c10::irange(size)) {
```
nit: Give idx and i more descriptive names

ezyang commented on torch/csrc/autograd/autograd_not_implemented_fallback.cpp:
```diff
@@ -0,0 +1,180 @@
+  _foreach_tensor([&](size_t _, size_t idx_arg, const at::Tensor& t) {
+    if (arguments[idx_arg].type()->kind() != c10::TypeKind::OptionalType) {
+      TORCH_CHECK(t.defined(), "Expected argument ", idx_arg, " of ", op_name, " to be defined.");
+    }
```
I don't feel like this test should be necessary. In none of the original VariableType kernels do we test if the Tensor inputs are defined or not. This is doing some sort of schema checking on the passed in call stack, but actually we should just assume that the stack inputs are well typed.

ezyang commented on torch/csrc/autograd/autograd_not_implemented_fallback.cpp:
```diff
@@ -0,0 +1,180 @@
+
+  #ifndef NDEBUG
+  // See NOTE [ TensorImpl and Storage Pointer Sanity Checks ]
+  std::vector<c10::IValue> stack_copy(*stack);
```
It's a bad idea to copy the entire stack, as you may be arbitrarily deep in the stack, and this would copy all of the parent frames as well. Just copy the argument segment of the stack.

ezyang commented on torch/csrc/autograd/autograd_not_implemented_fallback.cpp:
```diff
@@ -0,0 +1,180 @@
+  #ifndef NDEBUG
+  _foreach_tensor([&](size_t idx_tensor, size_t _, const at::Tensor& t) {
+    if (storage_saved.at(idx_tensor).has_value())
+      AT_ASSERT(storage_saved.at(idx_tensor).value().is_alias_of(t.storage()));
```
nit: use TORCH_INTERNAL_ASSERT

ezyang commented on test/test_autograd.py:
```diff
@@ -6039,6 +6041,246 @@ def run_functional_checks(test_case, test_name, name, apply_fn, run_grad_checks,
         test_case.assertEqual(self_variable.size(), self_variable.grad.size())
 
 
+# borrowed from test/test_cpp_extensions_jit.py
```
put this in a common helper file?

ezyang replied:
or, I'd actually suggest putting these tests in one of the extensions file, so that people can continue to run autograd tests without having to worry about compiling cpp extensions

ezyang reviewed (approved):
okey dokey

soulitzer committed:
Update on "Add autograd not implemented boxed fallback"


See description and discussion from https://github.com/pytorch/pytorch/pull/62450

[ghstack-poisoned]

soulitzer commented:
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30518572).

soulitzer committed:
Update on "Add autograd not implemented boxed fallback"


See description and discussion from https://github.com/pytorch/pytorch/pull/62450

Differential Revision: [D30518572](https://our.internmc.facebook.com/intern/diff/D30518572)

[ghstack-poisoned]

soulitzer committed:
Update on "Add autograd not implemented boxed fallback"


See description and discussion from https://github.com/pytorch/pytorch/pull/62450

Differential Revision: [D30518572](https://our.internmc.facebook.com/intern/diff/D30518572)

[ghstack-poisoned]

soulitzer committed:
Update on "Add autograd not implemented boxed fallback"


See description and discussion from https://github.com/pytorch/pytorch/pull/62450

Differential Revision: [D30518572](https://our.internmc.facebook.com/intern/diff/D30518572)

[ghstack-poisoned]

soulitzer commented:
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30518572).

soulitzer committed:
Update on "Add autograd not implemented boxed fallback"


See description and discussion from https://github.com/pytorch/pytorch/pull/62450

Differential Revision: [D30518572](https://our.internmc.facebook.com/intern/diff/D30518572)

[ghstack-poisoned]

soulitzer committed:
Update on "Add autograd not implemented boxed fallback"


See description and discussion from https://github.com/pytorch/pytorch/pull/62450

Differential Revision: [D30518572](https://our.internmc.facebook.com/intern/diff/D30518572)

[ghstack-poisoned]

soulitzer committed:
Update on "Add autograd not implemented boxed fallback"


See description and discussion from https://github.com/pytorch/pytorch/pull/62450

Differential Revision: [D30518572](https://our.internmc.facebook.com/intern/diff/D30518572)

[ghstack-poisoned]

soulitzer committed:
Update on "Add autograd not implemented boxed fallback"


See description and discussion from https://github.com/pytorch/pytorch/pull/62450

Differential Revision: [D30518572](https://our.internmc.facebook.com/intern/diff/D30518572)

[ghstack-poisoned]
